src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -344,133 +344,304 @@ public class SqlMutationEngine : IMutationEngine
-                // a DataApiBuilderException is thrown
-                catch (TransactionException)
+                if (context.OperationType is EntityActionOperation.Upsert || context.OperationType is EntityActionOperation.UpsertIncremental)
                 {
-                    throw _dabExceptionWithTransactionErrorMessage;
-                }
+                    DbResultSet? upsertOperationResult;
+                    DbResultSetRow? upsertOperationResultSetRow = null;
+                    JsonDocument? selectOperationResponse = null;
+                    bool isUpdateResultSet = false;
 
-                DbResultSetRow? dbResultSetRow = upsertOperationResult is not null ?
-                    (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
+                    try
+                    {
+                        // Creating an implicit transaction
+                        using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType())
+                        {
+                            upsertOperationResult = await PerformUpsertOperation(
+                                                                parameters,
+                                                                context);
 
-                if (upsertOperationResult is not null &&
-                    dbResultSetRow is not null && dbResultSetRow.Columns.Count > 0)
-                {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
+                            upsertOperationResultSetRow = upsertOperationResult is not null ?
+                        (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
 
-                    bool isUpdateResultSet = false;
-                    if (upsertOperationResult.ResultProperties.TryGetValue(IS_UPDATE_RESULT_SET, out object? isUpdateResultSetValue))
+                            if (upsertOperationResult is not null &&
+                                upsertOperationResultSetRow is not null &&
+                                upsertOperationResultSetRow.Columns.Count > 0 &&
+                                upsertOperationResult.ResultProperties.TryGetValue(IS_UPDATE_RESULT_SET, out object? isUpdateResultSetValue))
+                            {
+
+                                isUpdateResultSet = Convert.ToBoolean(isUpdateResultSetValue);
+                            }
+
+                            // The role with which the REST request is executed can have a database policy defined for the read action.
+                            // In such a case, to get the results back, a select query which honors the database policy is excuted.
+                            if (isDatabasePolicyDefinedForReadAction)
+                            {
+                                FindRequestContext findRequestContext = ConstructFindRequestContext(context, upsertOperationResultSetRow!, roleName);
```

Comment:
if `upsertOperationResult` is somehow null, then you can't guarantee that `upsertOperationResultSetRow` is non-null here.

Reply:
`upsertOperationResult` is never `null`. Methods such as `PerformUpsertOperation` `PerformMutationOperation` never actually return `null` values. However, all these methods are defined with a nullable return type and this forces us to write unnecessary null checks.

_Reason for defining these methods with nullable return types:_

All these methods internally use `_queryExecutor.ExecuteQueryAsync` which has a nullable return type. One of the parameters for `_queryExecutor.ExecuteQueryAsync` is a function delegate that contains logics related to the operation (in case of upsert, identifying whether an update or insert took place) and constructing some data structures with the results returned from the databse for further processing. 

Only the delegate used for query: `GetJsonResultAsync` returns `null` values. None of the delegates used for delete, upsert, insert, etc. return `null` values. 
To be able to support `GetJsonResultAsync`, `ExecuteQueryAsync` is defined with nullable return type value which forces us to write unneccesary `null` within `SqlMutationEngine`